feat: Bootstrap the mini-controller

yitsushi opened the pull request:
* There is a new field on Terraform: planOnly. It is not used anywhere,
  yet, tf controller ignores it.
* `webhook-receiver` is called `branch-based-planner` now, as the binary
  is not only the webhook receiver, this name fits more the
  functionality.
* The informer does nothing yet, it's just a bootstrap phase.
* The Informer and the Webhook receiver run in their own go routine, if
  any of them stops, it terminates the other as none of them are really
  useful without their parner.

Closes #588

References:
* https://github.com/weaveworks/tf-controller/issues/588

chanwit commented:
> The mini controller does nothing yet, it's just a bootstrap phase.

To simplify thing, we don't need any stuff like the reconciliation loop in the mini controller. 

The mini controller would be more like:
  - identifying target objects to be cloned, cloning set of objects per spec. the information would be obtained from the PR.
  - setting up an informer to watch Terraform CR objects with `planOnly=true` (you would also put labels during the object cloning procedure to make them query-able).

yitsushi commented:
> To simplify thing, we don't need any stuff like the reconciliation loop in the mini controller.

Something has to watch the resources to fetch the plan output and post it as a comment.

chanwit commented:
> Something has to watch the resources to fetch the plan output and post it as a comment, not?

exactly: so it's simpler to implement the controller like this with an informer:

> setting up an informer to watch Terraform CR objects with planOnly=true (you would also put labels during the object cloning procedure to make them query-able).

yitsushi commented:
Ah makes sense. Thank you for clarifying, my eyes skipped the "informer" part and recognised only the "don't need any stuff like the reconciliation loop".

yiannistri commented on cmd/branch-based-planner/main.go:
```diff
@@ -0,0 +1,130 @@
+	}(log.WithName("webhook-server"))
+
+	func(log logr.Logger) {
+		log.Info("Starting branch-blased planner informer")
```
```suggestion
		log.Info("Starting branch-based planner informer")
```

yiannistri commented on cmd/branch-based-planner/main.go:
```diff
@@ -0,0 +1,130 @@
+
+		// Does not matter if it was an error or not, if this routine is done for
+		// unknown reasons, stop the other routine too.
+		informerCancel()
```
```suggestion
		webhookCancel()
```

yitsushi replied:
That's intentional. If the informer stops, cancel webhook context, if the webhook stops, cancel informer context.

yiannistri replied:
Ah I didn't realise.

yiannistri commented on cmd/branch-based-planner/main.go:
```diff
@@ -0,0 +1,130 @@
+
+		// Does not matter if it was an error or not, if this routine is done for
+		// unknown reasons, stop the other routine too.
+		webhookCancel()
```
```suggestion
		informerCancel()
```

yitsushi replied:
That's intentional. If the informer stops, cancel webhook context, if the webhook stops, cancel informer context.

yiannistri commented on controllers/utils.go:
```diff
@@ -0,0 +1,27 @@
+	return func(o client.Object) []string {
+		terraform, ok := o.(*tfv1alpha2.Terraform)
+		if !ok {
+			panic(fmt.Sprintf("Expected a Kustomization, got %T", o))
```
```suggestion
			panic(fmt.Sprintf("Expected a Terraform object, got %T", o))
```

yiannistri commented on controllers/tf_bbp_controller.go:
Small typo `TerraformBranchBasesPlannerReconciler` -> `TerraformBranchBasedPlannerReconciler`

yitsushi replied:
Actually that whole block can be deleted as we don't have a controller anymore.

yiannistri reviewed (commented):
Some minor typos but generally looks fine 👍

yiannistri reviewed (dismissed):
Looks good 👍

chanwit commented on controllers/utils.go:
```diff
@@ -0,0 +1,27 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func IndexBy(kind string) func(o client.Object) []string {
+	return func(o client.Object) []string {
+		terraform, ok := o.(*tfv1alpha2.Terraform)
+		if !ok {
+			panic(fmt.Sprintf("Expected a Terraform object, got %T", o))
+		}
+
+		if terraform.Spec.SourceRef.Kind == kind {
+			namespace := terraform.GetNamespace()
+			if terraform.Spec.SourceRef.Namespace != "" {
+				namespace = terraform.Spec.SourceRef.Namespace
+			}
+			return []string{fmt.Sprintf("%s/%s", namespace, terraform.Spec.SourceRef.Name)}
+		}
+
+		return nil
+	}
+}
```
Please revert this too.

yitsushi replied:
done

yitsushi commented on cmd/branch-based-planner/main.go:
```diff
@@ -0,0 +1,175 @@
+	return nil
+}
+
+func setupCerts(opts *applicationOptions) (chan struct{}, *mtls.CertRotator) {
```
Yes it's a leftover from the controller.

yitsushi commented on controllers/tf_controller.go:
```diff
@@ -425,19 +425,19 @@ func shortRev(revision string) string {
 	// Index the Terraforms by the OCIRepository references they (may) point at.
 	if err := mgr.GetCache().IndexField(context.TODO(), &infrav1.Terraform{}, infrav1.OCIRepositoryIndexKey,
-		r.IndexBy(sourcev1b2.OCIRepositoryKind)); err != nil {
+		IndexBy(sourcev1b2.OCIRepositoryKind)); err != nil {
```
done

yitsushi commented:
@chanwit : removed unused (leftover from the Controller logic) code blocks and reverted the `IndexBy` function to be a function on the `TerraformReconciler` struct again, but fixed the error message @yiannistri spotted, so that's still a change in the `tf_controller.go` file.

chanwit reviewed (approved):
LGTM, thank you @yitsushi :1st_place_medal:
